Fixing rename refactoring for record components and compact constructors

lahodaj opened the pull request:
This is an attempt to fix rename refactoring for record components. Will still need some cleanup/more testing. Does not alter the AST shapes in any way, but uses them to perform the work. Will also rename "linked" elements for the component, like the accessor and the parameter for the (compact) constructor.

@mbien - not sure if you are looking at something similar? I apologize if I stepped on something you're looking at.

mbien reviewed (commented):
awesome! Looks like rename is working as long it is started via right click -> refactor -> rename.

the "realtime" rename mode (ctrl+R) will still produce wrong results. But this is probably linked to the semantic highlighting functionality which also doesn't associate record components with the compact constructor "params" for example.

The quickest fix for this might be to find the logic which decides if it should rename via dialog or in-place and let it open the dialog if it is a record component refactoring.

mbien commented on java/refactoring.java/src/org/netbeans/modules/refactoring/java/plugins/RenameRefactoringPlugin.java:
```diff
@@ -433,6 +438,27 @@ public void run(CompilationController info) throws Exception {
+                                            }
+                                        }
+                                    }
+                                }
```
L467 below would need the reverse case for renaming a record accessor which would have to be linked to the comonents i believe.

mbien reviewed (commented):
what doesn't work yet is:
 - renaming accessors (they are not linked to the components)
 - instant-renaming the record itself will not rename the compact constructor
 
The second point is caused due to `InstntRenamePerformer#computeChangePoints` looping through `ElementFilter.constructorsIn(el.getEnclosedElements()))`, however, this list does not include the compact constructor. In fact, it isn't even in` getEnclosedElements()`- so I don't know how to get it unfortunately.

```java
public class RenameRecord {

    private record Vec2d(int x, int y) {
        public Vec2d {
            if (x < 0 || y < 0) {
                throw new IllegalArgumentException();
            }
        }
        public Vec2d(int x) {
            this(x, 0);
        }
        @Override
        public int x() {
            return x;
        }
    }

    public static void main(String[] args) {

        Vec2d vec = new Vec2d(1);
        //       rename ^^^^^  (check compact constructor during instant rename)
        //                     (this does work if the record is made public)
        
        System.out.println(vec);
        System.out.println(vec.x);
        System.out.println(vec.x());
        //              rename ^  (doesn't rename component)
        System.out.println(vec.y);
        System.out.println(vec.y());
        //              rename ^  (NPE)
    }

}
```

mbien commented:
>   - instant-renaming the record itself will not rename the compact constructor
> 
> The second point is caused due to InstntRenamePerformer#computeChangePoints looping through ElementFilter.constructorsIn(el.getEnclosedElements())), however, this list does not include the compact constructor. In fact, it isn't even in getEnclosedElements()- so I don't know how to get it unfortunately.

ah I get it. The compact constructor is represented by the canonical constructor. `Utilities#getToken` can't convert it to a token though and returns null.
https://github.com/apache/netbeans/blob/36171a7e45f592c11be357fbef9a1dcf8d0e08ec/java/java.editor/src/org/netbeans/modules/java/editor/rename/InstantRenamePerformer.java#L377-L385

mbien commented on java/java.source.base/src/org/netbeans/api/java/source/ElementUtilities.java:
```diff
@@ -928,7 +929,143 @@ public ExecutableElement getDescriptorElement(TypeElement origin) {
+     *         that are linked to it by the Java record specification
+     * @since 2.70
+     */
+    public Collection<? extends Element> getLinkedRecordElements(Element forElement) {
```
great method!

mbien reviewed (approved):
code looks good and it fixes all rename scenarios I could come up with - awesome!

please don't forget to squash a little before merge

mbien commented:
spoke a bit too soon, using the example https://github.com/apache/netbeans/pull/7670#pullrequestreview-2243889285 and renaming a record component will rename it with instant-rename now which forgets about accessor methods. I believe last iteration used non-instant rename which works fine.

mbien commented on java/java.editor/src/org/netbeans/modules/java/editor/rename/InstantRenamePerformer.java:
```diff
@@ -450,14 +450,13 @@ private static boolean allowInstantRename(CompilationInfo info, Element e, Eleme
-            boolean isRecordComponent = e.getEnclosingElement().getKind() == ElementKind.RECORD;
-            if (isRecordComponent) {
-                return false;
-            }
```
might be still needed after all?

lahodaj replied:
I just placed the new check at a wrong place. Sorry for that.

mbien reviewed (approved):
perfect (works)

lahodaj commented:
I have squashed and fixed on more issue (the rename didn't work when the rename started from the constructor's parameter, because the `TreePathHandle` didn't use the `ElementHandle` for it, which then failed the refactorings in all other files expect the one that contains the parameter). If the tests are green, and there are no objections, I would merge.

dbalek reviewed (approved):
Looks fine. Thanks.

lahodaj commented:
I've found another problem yesterday: `ElementHandle.create` was happily creating `ElementHandle` for lambda parameters, but could not (quite obviously) resolve them back. In some cases, this then broke the refactoring. I've attempted to tweak that in this commit:
https://github.com/apache/netbeans/pull/7670/commits/56461a268907b628093ddf840cb875b329b02b79

mbien commented on java/java.source.base/apichanges.xml:
```diff
@@ -25,6 +25,18 @@
 <changes>
+    <change id="TreeMaker.RecordComponent">
+        <api name="javasource_base" />
+        <summary>Adding TreeMaker.RecordComponent</summary>
+        <version major="1" minor="2.70.0"/>
+        <date day="20" month="8" year="2023"/>
```
@lahodaj just wanting to make sure that 2023 was intended ;)

mbien commented:
still +1 from me - great fixes!
